feat: add `OptionalType` click parameter type

kimjinmyeong opened the pull request:
This PR resolves #1393 , and includes minor fixes to execute follow commands:
- `./backend.ai sesstpl create -f {template_filepath}`
- `./backend.ai session create-from-template {template_id}`
---

**Implement `OptionalType` class**

- `OptionalType` serves as a type wrapper that checks for `undefined` types.

**Revise to support for **YAML** format files and enable the creation of multiple templates.**
- To create multiple templates and iterate through a `for` loop when importing YAML file, the `safe_load` function should be replaced with the `safe_load_all` function. 
- To use the same code for creating query in both YAML and JSON formats, the JSON file should be modified as : 

<details>
  <summary>example_session_template.json</summary>

```json
[
  {
    "id": "c1b8441a-ba46-4a83-8727-de6645f521b4",
    "is_active": true,
    "domain_name": "default",
    "group_id": "2de2b969-1d04-48a6-af16-0bc8adb3c831",
    "user_uuid": "f38dea23-50fa-42a0-b5ae-338f5f4693f4",
    "type": "TASK",
    "name": "python_x86",
    "template": {
      "api_version": "6",
      "kind": "task_template",
      "metadata": {
        "name": "cr.backend.ai/multiarch/python",
        "tag": "3.10-ubuntu20.04"
      },
      "spec": {
        "session_type": "interactive",
        "kernel": {
          "image": "cr.backend.ai/multiarch/python:3.10-ubuntu20.04",
          "environ": {},
          "architecture": "x86_64",
          "run": null,
          "git": null
        },
        "scaling_group": "default",
        "mounts": {
        },
        "resources": {
          "cpu": "2",
          "mem": "4g"
        }
      }
    }
  },
  {
    "id": "59062449-4f57-4434-975d-add2a593438c",
    "is_active": true,
    "domain_name": "default",
    "group_id": "2de2b969-1d04-48a6-af16-0bc8adb3c831",
    "user_uuid": "f38dea23-50fa-42a0-b5ae-338f5f4693f4",
    "type": "TASK",
    "name": "python_arch64",
    "template": {
      "api_version": "6",
      "kind": "task_template",
      "metadata": {
        "name": "cr.backend.ai/multiarch/python",
        "tag": "3.10-ubuntu20.04"
      },
      "spec": {
        "session_type": "interactive",
        "kernel": {
          "image": "cr.backend.ai/multiarch/python:3.10-ubuntu20.04",
          "environ": {},
          "architecture": "aarch64",
          "run": null,
          "git": null
        },
        "scaling_group": "default",
        "mounts": {
        },
        "resources": {
          "cpu": "2",
          "mem": "4g"
        }
      }
    }
  }
]
```

</details>

<details>
  <summary>example_session_template.yaml</summary>

```yaml
---
id: c1b8441a-ba46-4a83-8727-de6645f521b4
is_active: true
domain_name: default
group_id: 2de2b969-1d04-48a6-af16-0bc8adb3c831
user_uuid: f38dea23-50fa-42a0-b5ae-338f5f4693f4
type: TASK
name: python_x86
template:
  api_version: '6'
  kind: task_template
  metadata:
    name: cr.backend.ai/multiarch/python
    tag: 3.10-ubuntu20.04
  spec:
    session_type: interactive
    kernel:
      image: cr.backend.ai/multiarch/python:3.10-ubuntu20.04
      environ: {}
      architecture: x86_64
      run: null
      git: null
    scaling_group: default
    mounts: {}
    resources:
      cpu: '2'
      mem: 4g
---
id: 59062449-4f57-4434-975d-add2a593438c
is_active: true
domain_name: default
group_id: 2de2b969-1d04-48a6-af16-0bc8adb3c831
user_uuid: f38dea23-50fa-42a0-b5ae-338f5f4693f4
type: TASK
name: python_arch64
template:
  api_version: '6'
  kind: task_template
  metadata:
    name: cr.backend.ai/multiarch/python
    tag: 3.10-ubuntu20.04
  spec:
    session_type: interactive
    kernel:
      image: cr.backend.ai/multiarch/python:3.10-ubuntu20.04
      environ: {}
      architecture: aarch64
      run: null
      git: null
    scaling_group: default
    mounts: {}
    resources:
      cpu: '2'
      mem: 4g

```

</details>

---

In `src/ai/backend/client/func/session_template.py`, I modifed `cls()` argument  to return the log with multiple `template_id` like this:
> Task template 868647ae08f94da38a2cf1a30bddb6e6, 76e7bed57f58487e934f5746dc5fdcb1 created and ready

**Minor fixes**
- revise the incorrect parameter order

**ai/backend/manager/utils.py**
```python
async def query_userinfo(
    conn: SAConnection,
    requester_uuid: UUID,
    requester_access_key: AccessKey,
    requester_role: UserRole,
```
**before ai/backend/manager/api/session.py**
```python
return await _query_userinfo(
            conn,
            request["user"]["uuid"],
            request["user"]["role"],
            request["keypair"]["access_key"],
```

- add the missing argument `callback_url` in `src/ai/backend/client/func/session.py`

**ai/backend/client/cli/session.py**
```python
        with Session() as session:
            try:
                compute_session = session.ComputeSession.create_from_template(
                    template_id,
                    image=image,
                    name=name,
                    type_=type_,
                    starts_at=starts_at,
                    enqueue_only=enqueue_only,
                    max_wait=max_wait,
                    no_reuse=no_reuse,
                    dependencies=depends,
                    callback_url=callback_url,
                    cluster_size=cluster_size,
```
**before ai/backend/client/func/session.py**
```python
async def create_from_template(
        cls,
        template_id: str,
        *,
        name: str | Undefined = undefined,
        type_: str | Undefined = undefined,
        starts_at: str = None,
        enqueue_only: bool | Undefined = undefined,
        max_wait: int | Undefined = undefined,
        # missing callback_url
        dependencies: Sequence[str] = None,  # cannot be stored in templates
        no_reuse: bool | Undefined = undefined,
```

fregataa commented on src/ai/backend/manager/api/session_template.py:
```diff
@@ -50,17 +50,18 @@ async def create(request: web.Request, params: Any) -> web.Response:
         except json.JSONDecodeError:
             try:
-                body = yaml.safe_load(params["payload"])
+                body = yaml.safe_load_all(params["payload"])
```
👍

fregataa commented on src/ai/backend/client/cli/params.py:
```diff
@@ -166,3 +168,14 @@ def convert(self, arg, param, ctx):
+class OptionalType(click.ParamType):
+    name = "Optional Type Wrapper"
+
+    def convert(self, value, param, ctx):
+        try:
+            if isinstance(value, (str, int)) or value == undefined:
+                return value
+        except ValueError as e:
+            self.fail(repr(e), param, ctx)
```
```suggestion
    def __init__(self, type_: type) -> None:
        super().__init__()
        self.type_ = type_

    def convert(self, value: Any, param, ctx):
        if isinstance(value, self.type_) or isinstance(value, undefined):
            return value
        self.fail(f"{value!r} is not valid `{self.type_}` or `undefined`", param, ctx)
```

Let `OptionalType` handle not only `str` and `int` types but also Any given type.
And it is good to pass detail failure messages.
Then, `OptionalType` can be used like below
```python
# example
@click.option("--name", type=OptionalType(str), default=undefined, help="Example option.")
```

fregataa reviewed (changes requested):
Thanks for your commitment!
I left some reviews

fregataa commented on src/ai/backend/client/cli/params.py:
```diff
@@ -166,3 +168,18 @@ def convert(self, arg, param, ctx):
+        super().__init__()
+        self.type_ = type_
+
+    def convert(self, value: Any, param, ctx):
+        try:
+            if isinstance(value, self.type_) or value is undefined:
+                return value
+        except ValueError:
+            self.fail(f"{value!r} is not valid `{self.type_}` or `undefined`", param, ctx)
```
```suggestion
    def convert(self, value: Any, param, ctx):
        try:
            if value is None or value is undefined:
                return value
            return self.type_(value)
        except ValueError:
            self.fail(f"{value!r} is not valid `{self.type_}` or `undefined`", param, ctx)
```
We should handle the case if `value` is not `type_` type instead of returning `None`.

kimjinmyeong commented:
@fregataa 
Here are some additional commits:
- In `OptionalType`, Since the `value` comes in as a `string` type, I added code to convert it to an `integer` type.
- Modified `architecture` and `sessionType` default values to `undefined` because the template value could be overridden by the default value.
- Added `dependencies` and `callbackURL` to the `params` as they were missing before.

fregataa commented on src/ai/backend/client/cli/params.py:
```diff
@@ -166,3 +168,26 @@ def convert(self, arg, param, ctx):
+            if value is None or value is undefined:
+                return value
+            else:
+                try:
+                    value = int(value)
+                    self.type_ = int
+                except ValueError:
+                    pass
```
```suggestion
                return self.type_(value)
```
`OptionalType(int)` cannot convert the input `str` value into `int`?
If not, we should think of general way to handle the value rather hard-parse it into `int`

kimjinmyeong replied:
> OptionalType(int) cannot convert the input str value into int?

Oh, it can! I have refactored it

fregataa reviewed (approved):
👍

fregataa commented:
it would be nice to embed an example `yaml` file to test session template!

kimjinmyeong commented:
@fregataa I'll make new PR to add example yaml template file soon!

kimjinmyeong commented on src/ai/backend/manager/api/session.py:
```diff
@@ -259,7 +259,6 @@ def check_and_return(self, value: Any) -> object:
         t.Key("startup_command", default=None): t.Null | t.String,
         t.Key("bootstrap_script", default=None): t.Null | t.String,
         t.Key("owner_access_key", default=None): t.Null | t.String,
-        tx.AliasedKey(["scaling_group", "scalingGroup"], default=None): t.Null | t.String,
```
We already check `"scaling_group`" here 

https://github.com/lablup/backend.ai/blob/9643a1471d2ea36da244f718cb241e2b196c5297/src/ai/backend/manager/api/session.py#L216

kimjinmyeong commented on src/ai/backend/manager/api/session.py:
```diff
@@ -383,10 +382,10 @@ async def _create(request: web.Request, params: dict[str, Any]) -> web.Response:
             tx.AliasedKey(["name", "session_name", "clientSessionToken"], default=undefined)
             >> "session_name": UndefChecker | t.Regexp(r"^(?=.{4,64}$)\w[\w.-]*\w$", re.ASCII),
             tx.AliasedKey(["image", "lang"], default=undefined): UndefChecker | t.Null | t.String,
-            tx.AliasedKey(["arch", "architecture"], default=DEFAULT_IMAGE_ARCH)
-            >> "architecture": t.String,
-            tx.AliasedKey(["type", "sessionType"], default="interactive")
-            >> "session_type": tx.Enum(SessionTypes),
```
when we create session template, these values already set default value here 
https://github.com/lablup/backend.ai/blob/4c62f7c38431d8f9fcdd52959b26f1d60b1ee05e/src/ai/backend/manager/models/session_template.py#L60-L65

kimjinmyeong commented on src/ai/backend/client/cli/session.py:
```diff
@@ -538,7 +550,8 @@ def _create_from_template_cmd(docs: str = None):
         "-g",
         "--group",
         metavar="GROUP_NAME",
-        default=None,
```
If a `None` value is passed, the `group` value gets overwritten with `None`, leading to an error.
>collected an error log [ERROR] "ai.backend.manager.api.exceptions.InvalidAPIParameters: Missing or invalid API parameters. (Invalid group)" from manager
